library/std/unstable_state_preparation.qs

```diff
@@ -0,0 +1,457 @@
+
+            if Length(coefficientsPadded) == 1 {
+                // Termination case
+                if AbsD(coefficientsPadded[0]) > tolerance {
+                    Exp([PauliZ], coefficientsPadded[0], [target]);
+                }
+            } else {
+                // Compute new coefficients.
+                let (coefficients0, coefficients1) = MultiplexZCoefficients(coefficientsPadded);
+                ApproximatelyMultiplexZ(tolerance, coefficients0, Most(control), target);
+                if AnyOutsideToleranceD(tolerance, coefficients1) {
+                    within {
+                        CNOT(Tail(control), target);
+                    } apply {
+                        ApproximatelyMultiplexZ(tolerance, coefficients1, Most(control), target);
+                    }
+                }
+            }
+        }
+
+        controlled (controlRegister, ...) {
+            // pad coefficients length to a power of 2.
+            let coefficientsPadded = Padded(2 ^ (Length(control) + 1), 0.0, Padded(-2 ^ Length(control), 0.0, coefficients));
+            let (coefficients0, coefficients1) = MultiplexZCoefficients(coefficientsPadded);
+            ApproximatelyMultiplexZ(tolerance,coefficients0, control, target);
+            if AnyOutsideToleranceD(tolerance,coefficients1) {
+                within {
+                    Controlled X(controlRegister, target);
+                } apply {
+                    ApproximatelyMultiplexZ(tolerance,coefficients1, control, target);
+                }
+            }
+        }
+    }
+
+    /// # Summary
+    /// Implementation step of multiply-controlled Z rotations.
+    internal function MultiplexZCoefficients(coefficients : Double[]) : (Double[], Double[]) {
+        let newCoefficientsLength = Length(coefficients) / 2;
+        mutable coefficients0 = [];
+        mutable coefficients1 = [];
+
+        for idxCoeff in 0 .. newCoefficientsLength - 1 {
+            set coefficients0 += [0.5 * (coefficients[idxCoeff] + coefficients[idxCoeff + newCoefficientsLength])];
+            set coefficients1 += [0.5 * (coefficients[idxCoeff] - coefficients[idxCoeff + newCoefficientsLength])];
+        }
+
+        return (coefficients0, coefficients1);
+    }
+
+    internal function AnyOutsideToleranceD(tolerance : Double, coefficients : Double[]) : Bool {
+        for coefficient in coefficients {
+            // NOTE: This function is not used in a recursion termination condition
+            // only to determine if the multiplex step needs to be applied.
+            // For tolerance 0.0 it is always applied due to >= comparison.
+            if AbsD(coefficient) >= tolerance {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    internal function AnyOutsideToleranceCP(tolerance : Double, coefficients : ComplexPolar[]) : Bool {
+        for coefficient in coefficients {
```

Comment:
It seems like the note/comment in `AnyOutsideToleranceD` also applies to this function so it might be worth also including it here if that's the case.

Reply:
This is not entirely correct. It is used as a recursion termination condition, but only in addition to the exhaustion of array elements. Another caveat is that this condition can technically become false even if tolerance is 0.0 (unlike the other one).
I added that comment to the other function because it was quite unexpected at first glance. (if that other function was used as the only recursion termination condition with tolerance 0.0, the recursion would never stop). This doesn't look like it's totally out of the ordinary so I'm not sure what comment would be appropriate here.